lib/Runtime/Library/JSONStringBuilder.cpp

```diff
@@ -0,0 +1,287 @@
+            }
+        }
+        this->EscapeAndAppendString(entry.propertyName);
+        this->AppendCharacter(_u(':'));
+        if (this->gap != nullptr)
+        {
+            this->AppendCharacter(_u(' '));
+        }
+
+        this->AppendJSONPropertyString(&entry);
+
+        isFirstMember = false;
+    }
+    NEXT_DLISTCOUNTED_ENTRY;
+
+    if (this->gap != nullptr)
+    {
+        this->AppendCharacter(_u('\n'));
+        this->AppendGap(stepbackLevel);
+    }
+
+    this->AppendCharacter(_u('}'));
+
+    this->indentLevel = stepbackLevel;
+}
+
+void
+JSONStringBuilder::AppendArrayString(_In_ JSONArray* valueArray)
+{
+    uint32 length = valueArray->length;
+    if (length == 0)
+    {
+        this->AppendCharacter(_u('['));
+        this->AppendCharacter(_u(']'));
+        return;
+    }
+
+    const uint32 stepbackLevel = this->indentLevel;
+    ++this->indentLevel;
+
+    JSONProperty* arr = valueArray->arr;
+
+    this->AppendCharacter(_u('['));
+
+    if (this->gap != nullptr)
+    {
+        this->AppendCharacter(_u('\n'));
+        this->AppendGap(indentLevel);
+    }
+
+    this->AppendJSONPropertyString(&arr[0]);
+
+    for (uint32 i = 1; i < length; ++i)
+    {
+        if (this->gap == nullptr)
+        {
+            this->AppendCharacter(_u(','));
+        }
+        else
+        {
+            this->AppendCharacter(_u(','));
+            this->AppendCharacter(_u('\n'));
+            this->AppendGap(indentLevel);
+        }
+        AppendJSONPropertyString(&arr[i]);
+    }
+
+    if (this->gap != nullptr)
+    {
+        this->AppendCharacter(_u('\n'));
+        this->AppendGap(stepbackLevel);
+    }
+    this->AppendCharacter(_u(']'));
+
+    this->indentLevel = stepbackLevel;
+}
+
+void
+JSONStringBuilder::AppendJSONPropertyString(_In_ JSONProperty* prop)
+{
+    switch (prop->type)
+    {
+    case JSONContentType::False:
+        this->AppendString(this->scriptContext->GetLibrary()->GetFalseDisplayString());
+        return;
+    case JSONContentType::True:
+        this->AppendString(this->scriptContext->GetLibrary()->GetTrueDisplayString());
+        return;
+    case JSONContentType::Null:
+        this->AppendString(this->scriptContext->GetLibrary()->GetNullDisplayString());
+        return;
+    case JSONContentType::Number:
+        this->AppendString(prop->numericValue.string);
+        return;
+    case JSONContentType::Object:
+        this->AppendObjectString(prop->obj);
+        return;
+    case JSONContentType::Array:
+        this->AppendArrayString(prop->arr);
+        return;
+    case JSONContentType::String:
+        this->EscapeAndAppendString(prop->stringValue);
+        return;
+    default:
+        Assume(UNREACHED);
+    }
+}
+
+void
+JSONStringBuilder::Build()
+{
+    this->AppendJSONPropertyString(this->jsonContent);
+    // Null terminate the string
+    *this->currentLocation = _u('\0');
+    Assert(this->currentLocation == buffer + bufferLength - 1);
```

Comment:
Is `bufferLength` only ever used in this assert? Should it be covered by an `#ifdef DEBUG` in that case?